File: java_skeleton/javabot/skeleton/Runner.java
package javabot.skeleton;

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.lang.Integer;
import java.lang.String;
import java.net.Socket;
import java.io.PrintWriter;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

/**
 * Interacts with the engine.
 */
public class Runner {
    private String host;
    private int port;
    private Bot pokerbot;
    private Socket socket;
    private PrintWriter outStream;
    private BufferedReader inStream;

    /**
     * Returns an incoming message from the engine.
     */
    public String[] receive() throws IOException {
        String line = this.inStream.readLine().trim();
        return line.split(" ");
    }

    /**
     * Encodes an action and sends it to the engine.
     */
    public void send(List<Action> actions) {
        String[] codes = new String[State.NUM_BOARDS];
        for (int i = 0; i < State.NUM_BOARDS; i++) {
            switch (actions.get(i).actionType) {
                case ASSIGN_ACTION_TYPE: {
                    codes[i] = (i + 1) + "A" + String.join(",", actions.get(i).cards);
                    break;
                }
                case FOLD_ACTION_TYPE: {
                    codes[i] = (i + 1) + "F";
                    break;
                }
                case CALL_ACTION_TYPE: {
                    codes[i] = (i + 1) + "C";
                    break;
                }
                case CHECK_ACTION_TYPE: {
                    codes[i] = (i + 1) + "K";
                    break;
                }
                default: {  // RAISE_ACTION_TYPE
                    codes[i] = (i + 1) + "R" + Integer.toString(actions.get(i).amount);
                    break;
                }
            }
        }
        String code = String.join(";", codes);
        this.outStream.println(code);
    }

    /**
     * Reconstructs the game tree based on the action history received from the engine.
     */
    public void run() throws IOException {
        GameState gameState = new GameState(0, 0, (float)0., 1);
        List<State> boardStates = new ArrayList<State>();
        for (int i = 0; i < State.NUM_BOARDS; i++) {
            boardStates.add(new BoardState((i+1)*State.BIG_BLIND, Arrays.asList(0, 0),
                                            Arrays.asList(Arrays.asList(""), Arrays.asList("")),
                                            Arrays.asList(""), null));
        }
        State roundState = new RoundState(-2, 0, Arrays.asList(0, 0),
                                          Arrays.asList(Arrays.asList(""), Arrays.asList("")),
                                          boardStates, null);
        int active = 0;
        boolean roundFlag = true;
        while (true) {
            String[] packet = this.receive();
            for (String clause : packet) {
                String leftover = clause.substring(1, clause.length());
                switch (clause.charAt(0)) {
                    case 'T': {
                        gameState = new GameState(gameState.bankroll, gameState.oppBankroll, Float.parseFloat(leftover), gameState.roundNum);
                        break;
                    }
                    case 'P': {
                        active = Integer.parseInt(leftover);
                        break;
                    }
                    case 'H': {
                        String[] cards = leftover.split(",");
                        List<List<String>> hands = new ArrayList<List<String>>(
                            Arrays.asList(
                                new ArrayList<String>(),
                                new ArrayList<String>()
                            )
                        );
                        hands.set(active, Arrays.asList(cards));
                        String[] oppHands = new String[cards.length];
                        Arrays.fill(oppHands, "");
                        hands.set(1 - active, Arrays.asList(oppHands));
                        List<String> deck = new ArrayList<String>(Arrays.asList("", "", "", "", ""));
                        List<Integer> pips = Arrays.asList(State.SMALL_BLIND, State.BIG_BLIND);
                        boardStates = new ArrayList<State>();
                        for (int i = 0; i < State.NUM_BOARDS; i++) {
                            boardStates.add(new BoardState((i+1)*State.BIG_BLIND, pips,
                                            Arrays.asList(new ArrayList<String>(), new ArrayList<String>()),
                                            deck, null));
                        }
                        List<Integer> stacks = Arrays.asList(State.STARTING_STACK - State.NUM_BOARDS*State.SMALL_BLIND,
                                                             State.STARTING_STACK - State.NUM_BOARDS*State.BIG_BLIND);
                        roundState = new RoundState(-2, 0, stacks, hands, boardStates, null);
                        if (roundFlag) {
                            this.pokerbot.handleNewRound(gameState, (RoundState)roundState, active);
                            roundFlag = false;
                        }
                        break;
                    }
                    case 'D': {
                        String[] subclauses = clause.split(";");
                        int delta = Integer.parseInt(subclauses[0].substring(1, subclauses[0].length()));
                        int oppDelta = Integer.parseInt(subclauses[1].substring(1, subclauses[1].length()));
                        List<Integer> deltas = new ArrayList<Integer>(Arrays.asList(delta, oppDelta));
                        deltas.set(active, delta);
                        deltas.set(1-active, oppDelta);
                        roundState = new TerminalState(deltas, ((TerminalState)roundState).previousState);
                        gameState = new GameState(gameState.bankroll + delta, gameState.oppBankroll + oppDelta, gameState.gameClock, gameState.roundNum);
                        this.pokerbot.handleRoundOver(gameState, (TerminalState)roundState, active);
                        gameState = new GameState(gameState.bankroll, gameState.oppBankroll, gameState.gameClock, gameState.roundNum + 1);
                        roundFlag = true;
                        break;
                    }
                    case 'Q': {
                        return;
                    }
                    case '1': {
                        roundState = this.parseMultiCode(clause, roundState, active);
                        break;
                    }
                    default: {
                        break;
                    }
                }
            }
            if (roundFlag) {  // ack the engine
                List<Action> ack = new ArrayList<Action>();
                for (int i = 0; i < State.NUM_BOARDS; i++) {
                    ack.add(new Action(ActionType.CHECK_ACTION_TYPE));
                }
                this.send(ack);
            } else {
                List<Action> actions = this.pokerbot.getActions(gameState, (RoundState)roundState, active);
                this.send(actions);
            }
        }
    }

    /**
     * Parses clauses which contain codes across multiple boards.
      */
    public State parseMultiCode(String clause, State roundState, int active) {
        String[] subclauses = clause.split(";");
        if (clause.contains("B")) {
            List<State> newBoardStates = new ArrayList<State>();
            for (int i = 0; i < State.NUM_BOARDS; i++) {
                String leftover = subclauses[i].substring(2, subclauses[i].length());
                String[] cards = leftover.split(",");
                List<String> revisedDeck = new ArrayList<String>(Arrays.asList("", "", "", "", ""));
                for (int j = 0; j < cards.length; j++) {
                    revisedDeck.set(j, cards[j]);
                }
                if (((RoundState)roundState).boardStates.get(i) instanceof BoardState) {
                    BoardState maker = (BoardState)((RoundState)roundState).boardStates.get(i);
                    newBoardStates.add(new BoardState(maker.pot, maker.pips, maker.hands,
                                                        revisedDeck, maker.previousState));
                } else {
                    TerminalState terminal = (TerminalState)((RoundState)roundState).boardStates.get(i);
                    BoardState maker = (BoardState)terminal.previousState;
                    newBoardStates.add(new TerminalState(terminal.deltas,
                                                        new BoardState(maker.pot, maker.pips, maker.hands,
                                                                        revisedDeck, maker.previousState,
                                                                        maker.settled)));
                }
            }
            RoundState maker = (RoundState)roundState;
            return new RoundState(maker.button, maker.street, maker.stacks, maker.hands,
                                    newBoardStates, maker.previousState);
        } else if (clause.contains("O")) {
            List<State> newBoardStates = new ArrayList<State>();
            roundState = (RoundState)((TerminalState)roundState).previousState;
            for (int i = 0; i < State.NUM_BOARDS; i++) {
                String leftover = subclauses[i].substring(2, subclauses[i].length());
                if ("".equals(leftover)) {
                    newBoardStates.add(((RoundState)roundState).boardStates.get(i));
                } else {
                    // backtrack
                    String[] cards = leftover.split(",");
                    TerminalState terminal = (TerminalState)((RoundState)roundState).boardStates.get(i);
                    BoardState maker = (BoardState)terminal.previousState;
                    List<List<String>> revisedHands = new ArrayList<List<String>>(maker.hands);
                    revisedHands.set(1 - active, Arrays.asList(cards[0], cards[1]));
                    newBoardStates.add(new TerminalState(terminal.deltas, new BoardState(maker.pot, maker.pips, revisedHands, maker.deck, maker.previousState, maker.settled)));
                }
            }
            RoundState maker = (RoundState)roundState;
            roundState = new RoundState(maker.button, maker.street, maker.stacks, maker.hands, newBoardStates, maker.previousState);
            return new TerminalState(Arrays.asList(0, 0), roundState);
        }
        else {
            List<Action> actions = new ArrayList<Action>();
            for(String subclause : subclauses) {
                String leftover = subclause.substring(2, subclause.length());
                switch (subclause.charAt(1)) {
                    case 'F': {
                        actions.add(new Action(ActionType.FOLD_ACTION_TYPE));
                        break;
                    }
                    case 'C': {
                        actions.add(new Action(ActionType.CALL_ACTION_TYPE));
                        break;
                    }
                    case 'K': {
                        actions.add(new Action(ActionType.CHECK_ACTION_TYPE));
                        break;
                    }
                    case 'R': {
                        actions.add(new Action(ActionType.RAISE_ACTION_TYPE, Integer.parseInt(leftover)));
                        break;
                    }
                    case 'A': {
                        String[] cards = leftover.split(",");
                        if ("".equals(leftover)) {
                            actions.add(new Action(ActionType.ASSIGN_ACTION_TYPE, Arrays.asList("", "")));
                        } else {
                            actions.add(new Action(ActionType.ASSIGN_ACTION_TYPE, Arrays.asList(cards)));
                        }
                        break;
                    }
                    default: {
                        break;
                    }
                }
            }
            return ((RoundState)roundState).proceed(actions);
        }
    }

    /**
     * Parses arguments corresponding to socket connection information.
     */
    public void parseArgs(String[] rawArgs) {
        boolean hostFlag = false;
        this.host = "localhost";
        for (String arg : rawArgs) {
            if (arg.equals("-h") | arg.equals("--host")) {
                hostFlag = true;
            } else if (arg.equals("--port")) {
                // nothing to do
            } else if (hostFlag) {
                this.host = arg;
                hostFlag = false;
            } else {
                this.port = Integer.parseInt(arg);
            }
        }
    }

    /**
     * Runs the pokerbot.
     */
    public void runBot(Bot pokerbot) {
        this.pokerbot = pokerbot;
        try {
            this.socket = new Socket(this.host, this.port);
            this.socket.setTcpNoDelay(true);
            this.outStream = new PrintWriter(socket.getOutputStream(), true);
            this.inStream = new BufferedReader(new InputStreamReader(socket.getInputStream()));
        } catch (IOException e) {
            System.out.println("Could not connect to " + host + ":" + Integer.toString(port));
            return;
        }
        try {
            this.run();
            this.inStream.close();
            this.inStream = null;
            this.outStream.close();
            this.outStream = null;
            this.socket.close();
            this.socket = null;
        } catch (IOException e) {
            System.out.println("Engine disconnected.");
        }
    }
}

File: python_skeleton/skeleton/runner.py
'''
The infrastructure for interacting with the engine.
'''
import argparse
import socket
from .actions import FoldAction, CallAction, CheckAction, RaiseAction, AssignAction
from .states import GameState, TerminalState, RoundState, BoardState
from .states import STARTING_STACK, BIG_BLIND, SMALL_BLIND, NUM_BOARDS
from .bot import Bot


class Runner():
    '''
    Interacts with the engine.
    '''

    def __init__(self, pokerbot, socketfile):
        self.pokerbot = pokerbot
        self.socketfile = socketfile

    def receive(self):
        '''
        Generator for incoming messages from the engine.
        '''
        while True:
            packet = self.socketfile.readline().strip().split(' ')
            if not packet:
                break
            yield packet

    def send(self, actions):
        '''
        Encodes actions and sends it to the engine.
        '''
        codes = [''] * NUM_BOARDS
        for i in range(NUM_BOARDS):
            if isinstance(actions[i], AssignAction):
                codes[i] = str(i+1) + 'A' + ','.join(actions[i].cards)
            elif isinstance(actions[i], FoldAction):
                codes[i] = str(i+1) + 'F'
            elif isinstance(actions[i], CallAction):
                codes[i] = str(i+1) + 'C'
            elif isinstance(actions[i], CheckAction):
                codes[i] = str(i+1) + 'K'
            else:  # isinstance(action, RaiseAction)
                codes[i] = str(i+1) + 'R' + str(actions[i].amount)
        code = ';'.join(codes)
        self.socketfile.write(code + '\n')
        self.socketfile.flush()
       
    def run(self):
        '''
        Reconstructs the game tree based on the action history received from the engine.
        '''
        game_state = GameState(0, 0, 0., 1)
        round_state = None
        active = 0
        round_flag = True
        for packet in self.receive():
            for clause in packet:
                if clause[0] == 'T':
                    game_state = GameState(game_state.bankroll, game_state.opp_bankroll, float(clause[1:]), game_state.round_num)
                elif clause[0] == 'P':
                    active = int(clause[1:])
                elif clause[0] == 'H':
                    cards = clause[1:].split(',')
                    hands = [[], []]
                    hands[active] = cards
                    hands[1-active] = ['']*(2*NUM_BOARDS)
                    deck = ["", "", "", "", ""]
                    pips = [SMALL_BLIND, BIG_BLIND]
                    board_states = [BoardState((i+1)*BIG_BLIND, pips, [[]]*2, deck, None) for i in range(NUM_BOARDS)]
                    stacks = [STARTING_STACK - NUM_BOARDS*SMALL_BLIND, STARTING_STACK - NUM_BOARDS*BIG_BLIND]
                    round_state = RoundState(-2, 0, stacks, hands, board_states, None)
                    if round_flag:
                        self.pokerbot.handle_new_round(game_state, round_state, active)
                        round_flag = False
                elif clause[0] == 'D':
                    assert isinstance(round_state, TerminalState)
                    subclauses = clause.split(';')
                    delta = int(subclauses[0][1:])
                    opp_delta = int(subclauses[1][1:])
                    deltas = [delta, opp_delta]
                    deltas[active] = delta
                    deltas[1-active] = opp_delta
                    round_state = TerminalState(deltas, round_state.previous_state)
                    game_state = GameState(game_state.bankroll + delta, game_state.opp_bankroll + opp_delta, game_state.game_clock, game_state.round_num)
                    self.pokerbot.handle_round_over(game_state, round_state, active)
                    game_state = GameState(game_state.bankroll, game_state.opp_bankroll, game_state.game_clock, game_state.round_num + 1)
                    round_flag = True
                elif clause[0] == 'Q':
                    return
                elif clause[0] == '1':
                    round_state = parse_multi_code(clause, round_state, active)
            if round_flag:  # ack the engine
                self.send([CheckAction()]*NUM_BOARDS)
            else:
                assert active == round_state.button % 2
                actions = self.pokerbot.get_actions(game_state, round_state, active)
                self.send(actions)


def parse_multi_code(clause, round_state, active):
    subclauses = clause.split(';')
    if 'B' in clause:
        new_board_states = [None] * NUM_BOARDS
        for i in range(NUM_BOARDS):
            leftover = subclauses[i][2:]
            cards = leftover.split(',')
            revised_deck = ["", "", "", "", ""]
            for j in range(len(cards)):
                revised_deck[j] = cards[j]
            if isinstance(round_state.board_states[i], BoardState):
                maker = round_state.board_states[i]
                new_board_states[i] = BoardState(maker.pot, maker.pips, maker.hands, revised_deck, maker.previous_state)
            else:
                terminal = round_state.board_states[i]
                maker = terminal.previous_state
                new_board_states[i] = TerminalState(terminal.deltas, BoardState(maker.pot, maker.pips, maker.hands, revised_deck, maker.previous_state, maker.settled))
        return RoundState(round_state.button, round_state.street, round_state.stacks, round_state.hands, new_board_states, round_state.previous_state)
    elif 'O' in clause:
        new_board_states = [None] * NUM_BOARDS
        round_state = round_state.previous_state
        for i in range(NUM_BOARDS):
            leftover = subclauses[i][2:]
            if leftover == "":
                new_board_states[i] = round_state.board_states[i]
            else:
                cards = leftover.split(',')
                terminal = round_state.board_states[i]
                maker = terminal.previous_state
                revised_hands = maker.hands
                revised_hands[1-active] = cards
                new_board_states[i] = TerminalState(terminal.deltas, BoardState(maker.pot, maker.pips, revised_hands, maker.deck, maker.previous_state, maker.settled))
        round_state = RoundState(round_state.button, round_state.street, round_state.stacks, round_state.hands, new_board_states, round_state.previous_state)
        return TerminalState([0, 0], round_state)
    else:
        actions = [None] * NUM_BOARDS
        for i in range(NUM_BOARDS):
            subclause = subclauses[i]
            leftover = subclause[2:]
            if subclause[1] == 'F':
                actions[i] = FoldAction()
            elif subclause[1] == 'C':
                actions[i] = CallAction()
            elif subclause[1] == 'K':
                actions[i] = CheckAction()
            elif subclause[1] == 'R':
                actions[i] = RaiseAction(int(leftover))
            elif subclause[1] == 'A':
                cards = leftover.split(',')
                if leftover == "":
                    actions[i] = AssignAction(["", ""])
                else:
                    actions[i] = AssignAction(cards)
        return round_state.proceed(actions)

def parse_args():
    '''
    Parses arguments corresponding to socket connection information.
    '''
    parser = argparse.ArgumentParser(prog='python3 player.py')
    parser.add_argument('--host', type=str, default='localhost', help='Host to connect to, defaults to localhost')
    parser.add_argument('port', type=int, help='Port on host to connect to')
    return parser.parse_args()

def run_bot(pokerbot, args):
    '''
    Runs the pokerbot.
    '''
    assert isinstance(pokerbot, Bot)
    try:
        sock = socket.create_connection((args.host, args.port))
    except OSError:
        print('Could not connect to {}:{}'.format(args.host, args.port))
        return
    socketfile = sock.makefile('rw')
    runner = Runner(pokerbot, socketfile)
    runner.run()
    socketfile.close()
    sock.close()
